File: Ruby/persistent_bugger.rb
# Write a function, persistence, that takes in a positive parameter num and returns its multiplicative persistence, which is the number of times you must multiply the digits in num until you reach a single digit.

# For example (Input --> Output):
# 39 --> 3 (because 3*9 = 27, 2*7 = 14, 1*4 = 4 and 4 has only one digit)
# 999 --> 4 (because 9*9*9 = 729, 7*2*9 = 126, 1*2*6 = 12, and finally 1*2 = 2)
# 4 --> 0 (because 4 is already a one-digit number)

# SOLUTION ONE
@count = 0

def persistence(n)
  if n < 10
    return p @count
  else
    arr = n.to_s.split(//)
    res = 1
    for i in arr
      res = res * i.to_i
    end
    @count += 1
    persistence(res)
  end
end

persistence(39)

# SOLUTION TWO
def persistence(n)
  recursive(n, 0)
end

def recursive(n, iteration)
  if n < 10
    return p iteration
  else
    arr = n.to_s.split(//)
    res = 1
    for i in arr
      res = res * i.to_i
    end
    recursive(res, iteration += 1)
  end
end

persistence(39)

File: Python/previous_exercises/persistent_bugger.py
# # SOLUTION ONE
# def persistence(n, t=0):
#     s = str(n)
#     num = 1
#     if n < 10:
#         print(t)
#         print(type(t))
#         return t
#     for c in s:
#         num *= int(c)
#     persistence(num, t + 1)


# # SOLUTION TWO
# def persistence(n):
#     recursive(n, 0)


# def recursive(n, t):
#     s = str(n)
#     num = 1
#     if n < 10:
#         print(t)
#         return t
#     for c in s:
#         num *= int(c)
#     recursive(num, t + 1)

# SOLUTION THREE
def persistence(n):
    n = str(n)
    count = 0
    while len(n) > 1:
        p = 1
        for i in n:
            p *= int(i)
        n = str(p)
        count += 1
    return count


persistence(39)
persistence(4)
persistence(25)
persistence(999)
